tools/shield_utils.py

```diff
@@ -0,0 +1,427 @@
+    if match is None:
+        log.die("Invalid shield: '{}'. Please stick with the pattern:"
+                " '{{NUMA_NODE_ID}};{{CPUS_COUNT}}'.".format(shield),
+                ExitCode.EINVPATTERN)
+
+    numa_node_id = int(match.group(1))
+    cpus_count = int(match.group(2))
+    if not _is_node_online(numa_node_id):
+        log.die("NUMA node {} is not online.".format(numa_node_id),
+                ExitCode.EOFFNODE)
+
+    if len(_get_available_cpus(numa_node_id)) < cpus_count:
+        log.die("Can not provide {} CPUs for NUMA node {}."
+                .format(cpus_count, numa_node_id), ExitCode.ENOAVAILCPU)
+
+    if not _is_memory_node_available(numa_node_id):
+        log.die("NUMA node {} memory is assigned to an existing cpuset,"
+                " other than '/' and 'docker'.".format(numa_node_id),
+                ExitCode.ENOAVAILMEM)
+
+
+def ensure_environment(numa_node_id, cpus_count, return_to_stdout=False):
+    """Exit if NUMA node id or cpus count CPU cores are not free.
+
+    It verifies if the environment can be satisfied by using the
+    specified numa node id and cpus count of the same numa node.
+    Exits if it can not produce a list of cpus count CPU cores, of the
+    numa node id.
+
+    Args:
+        * numa_node_id (positive int).
+        * cpus_count (positive int).
+        * return_to_stdout (boolean): False by default. If set to True,
+        it changes the format of the return type.
+
+    Returns:
+        * list: A list of positive ints with all the CPUs cores that can
+                be used to set up an isolated enviornment for numa node id,
+                which is not used by other cpusets.
+        * str: Returns a cpuset list formatted set of CPUs cores which can be
+               written directly into the `cpuset.cpus` file of a cpuset.
+    """
+    if numa_node_id < 0:
+        log.die("Invalid NUMA node id: {}.".format(numa_node_id),
+                ExitCode.EINVARG)
+
+    if cpus_count < 1:
+        log.die("Invalid cpus count: {}.".format(cpus_count), ExitCode.EINVARG)
+
+    if not _is_node_online(numa_node_id):
+        log.die("NUMA node is not online: {}.".format(numa_node_id),
+                ExitCode.EOFFNODE)
+
+    if not _is_memory_node_available(numa_node_id):
+        log.die("NUMA node {} memory is assigned to other cpuset."
+                .format(numa_node_id), ExitCode.ENOAVAILMEM)
+
+    cpu_list = _get_cpulist(numa_node_id)
+    avail_cpus = _get_available_cpus(numa_node_id)
+
+    cpu_list_len = functools.reduce(lambda acc, cpu: acc + cpu, cpu_list, 0)
+
+    if cpu_list_len < cpus_count or len(avail_cpus) < cpus_count:
+        log.die("Can not provide {} CPUs for NUMA node {}.".
+                format(cpus_count, numa_node_id), ExitCode.ENOAVAILCPU)
+
+    if return_to_stdout:
+        return ','.join(map(str, avail_cpus))
+
+    return avail_cpus
+
+
+def create_system_cpuset(excluded_numa_node_id):
```

Comment:
This is very similar to what `cset shield` does. Could we try to use it instead ?

Reply:
Yes, indeed. This PR was inspired by how the tool is designed. Please see my reasoning above about why it is difficult to use `cset shield` in this case.